server/src/main/java/org/opensearch/indices/replication/SegmentReplicator.java

```diff
@@ -102,6 +111,135 @@ SegmentReplicationTarget startReplication(
         return target;
     }
 
+    /**
+     * Retrieves segment replication statistics for a specific shard.
+     * Its computed based on the last and first entry in the replicationCheckpointStats map.
+     * The Last entry gives the Bytes behind, and the difference in the first and last entry provides the lag.
+     *
+     * @param shardId The shardId to get statistics for
+     * @return ReplicationStats containing bytes behind and replication lag information
+     */
+    public ReplicationStats getSegmentReplicationStats(final ShardId shardId) {
+        final ConcurrentNavigableMap<Long, ReplicationCheckpointStats> existingCheckpointStats = replicationCheckpointStats.get(shardId);
+        if (existingCheckpointStats == null || existingCheckpointStats.isEmpty()) {
+            return ReplicationStats.empty();
+        }
+
+        Map.Entry<Long, ReplicationCheckpointStats> lowestEntry = existingCheckpointStats.firstEntry();
+        Map.Entry<Long, ReplicationCheckpointStats> highestEntry = existingCheckpointStats.lastEntry();
+
+        long bytesBehind = highestEntry.getValue().getBytesBehind();
+        long replicationLag = bytesBehind > 0L
+            ? TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - lowestEntry.getValue().getTimestamp())
+            : 0;
+
+        return new ReplicationStats(bytesBehind, bytesBehind, replicationLag);
+    }
+
+    /**
+     * Updates the latest checkpoint of the primary for the replica shard and then
+     * calculates checkpoint statistics for the replica shard with the latest checkpoint information.
+     * This method maintains statistics about how far behind replica shards are from the primary.
+     * It calculates the bytes behind by comparing the latest-received and current checkpoint in the indexShard,
+     * and it maintains the bytes behind and timestamp for each segmentInfosVersion of latestCheckPoint.
+     * <pre>
+     * Example:
+     * {
+     *     [replica][0] : {
+     *                       7 : {bytesBehind=0, timestamp=1700220000000}
+     *                       8 : {bytesBehind=100, timestamp=1700330000000}
+     *                       9 : {bytesBehind=150, timestamp=1700440000000}
+     *                    }
+     * }
+     * </pre>
+     * @param latestReceivedCheckPoint The most recent checkpoint from the primary
+     * @param indexShard The index shard where its updated
+     */
+    public void updateReplicationCheckpointStats(final ReplicationCheckpoint latestReceivedCheckPoint, final IndexShard indexShard) {
+        ReplicationCheckpoint latestPrimaryCheckPoint = this.latestPrimaryCheckpoint.get(indexShard.shardId());
```

Comment:
nit - pls name this something else for readability, its the same name as your map.
```suggestion
        ReplicationCheckpoint primaryCheckpoint = this.latestPrimaryCheckpoint.get(indexShard.shardId());
```

Reply:
Done